input/tar2/tar_input.go

```diff
@@ -0,0 +1,163 @@
+package tar2
+
+import (
+	"archive/tar"
+	"bytes"
+	"crypto/sha512"
+	"encoding/base64"
+	"hash"
+	"io"
+	"os"
+	"path"
+
+	"github.com/spacemonkeygo/errors"
+	"github.com/spacemonkeygo/errors/try"
+	"polydawn.net/repeatr/def"
+	"polydawn.net/repeatr/input"
+	"polydawn.net/repeatr/lib/fs"
+	"polydawn.net/repeatr/lib/fshash"
+	"polydawn.net/repeatr/lib/treewalk"
+)
+
+const Type = "tar"
+
+var _ input.Input = &Input{} // interface assertion
+
+type Input struct {
+	spec          def.Input
+	hasherFactory func() hash.Hash
+}
+
+func New(spec def.Input) input.Input {
+	if spec.Type != Type {
+		panic(errors.ProgrammerError.New("This input implementation supports definitions of type %q, not %q", Type, spec.Type))
+	}
+	_, err := os.Stat(spec.URI)
+	if os.IsNotExist(err) {
+		panic(input.DataSourceUnavailableError.New("Input URI %q must be a tar file", spec.URI))
+	}
+	return &Input{
+		spec:          spec,
+		hasherFactory: sha512.New384,
+	}
+}
+
+func (i Input) Apply(destinationRoot string) <-chan error {
+	done := make(chan error)
+	go func() {
+		defer close(done)
+		try.Do(func() {
+			// do make a dir for untaring into.
+			// tars may specify permission and time bits for their top dir, but if not, we'll start with sane defaults.
+			err := os.MkdirAll(destinationRoot, 0755)
+			if err != nil {
+				panic(input.TargetFilesystemUnavailableIOError(err))
+			}
+			// daemon uid and gid are fine for now (they're always 0);
+			// forcing a constant time happens along with all the other dirs, since
+			// the bucket normalizes to include a root attribute set if the tar doesn't have one.
+
+			// open the tar file; preparing decompression as necessary
+			file, err := os.OpenFile(i.spec.URI, os.O_RDONLY, 0755)
+			if err != nil {
+				panic(input.DataSourceUnavailableIOError(err))
+			}
+			defer file.Close()
+			reader, err := Decompress(file)
+			if err != nil {
+				panic(input.DataSourceUnavailableError.New("could not start decompressing: %s", err))
+			}
+			tarReader := tar.NewReader(reader)
+
+			// unroll the tar, copying and accumulating data for integrity check
+			bucket := &fshash.MemoryBucket{}
+			walk(tarReader, destinationRoot, bucket, i.hasherFactory)
+
+			// hash whole tree
+			actualTreeHash, _ := fshash.Hash(bucket, i.hasherFactory)
+
+			// verify total integrity
+			expectedTreeHash, err := base64.URLEncoding.DecodeString(i.spec.Hash)
```

Comment:
I don't think there is one.  https://github.com/jbenet/multihash is an interesting attempt and maybe we should consider using it.

But for now I still want to explicitly handwave that we're not migration-proofed on this front.  The hard part in this scenario isn't actually including a versioning bit; the semantics require a lot of thought.  (I.e. since changing the hash is the equivalent of a "major version" bump in semver terms, compatibility during a transition is... limited, essentially requiring multiple full "messages" in both relevant formats.  And that... you can see how that gets nontrivial fast.)  We should tackle it someday, but tl;dr I expect it's going to be a very big line-item on its own.